src/Core/Services/Implementations/AzureQueueService.cs

```diff
@@ -32,36 +34,62 @@ public async Task CreateManyAsync(IEnumerable<T> messages)
                 return;
             }
 
-            foreach (var json in SerializeMany(messages))
+            if (!messages.Skip(1).Any())
+            {
+                await CreateAsync(messages.First());
+                return;
+            }
+
+            foreach (var json in SerializeMany(messages, _jsonSettings))
             {
                 await _queueClient.SendMessageAsync(json);
             }
         }
 
-
-        private IEnumerable<string> SerializeMany(IEnumerable<T> messages)
+        protected IEnumerable<string> SerializeMany(IEnumerable<T> messages, JsonSerializerSettings jsonSettings)
         {
-            string SerializeMessage(T message) => JsonConvert.SerializeObject(message, _jsonSettings);
+            // Calculate Base-64 encoded text with padding
+            int getBase64Size(int byteCount) => ((4 * byteCount / 3) + 3) & ~3;
 
-            var messagesLists = new List<List<T>> { new List<T>() };
-            var strings = new List<string>();
-            var ListMessageLength = 2; // to account for json array brackets "[]"
-            foreach (var (message, jsonEvent) in messages.Select(m => (m, SerializeMessage(m))))
-            {
+            var messagesList = new List<string>();
+            var messagesListSize = 0;
+            
+            int calculateByteSize(int totalSize, int toAdd) =>
```

Comment:
Haven't looked into it in-depth. But there is a `Base64.GetMaxEncodedToUtf8Length` helper we could perhaps use instead? https://docs.microsoft.com/en-us/dotnet/api/system.buffers.text.base64.getmaxencodedtoutf8length?view=net-5.0#system-buffers-text-base64-getmaxencodedtoutf8length(system-int32)

Reply:
Hi @Hinton , I missed that, however from the [src of that method](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Memory/src/System/Buffers/Text/Base64Encoder.cs#L155) it looks like my implementation here is more accurate and most of the logic is still necessary, including this particular method that determines the total combined "next" JSON block to be added to see if it needs to start a new batch. I think we're good on this one IMHO.